charlotte_core/src/arch/x86_64/exceptions/exceptions.rs

```diff
@@ -0,0 +1,120 @@
+#![feature(asm)]
+
+static mut BSP_REGS: [u64; 16] = [0; 16];
+
+pub fn save_regs() {
+    unsafe {
+        asm!(
+            "mov [{} + 0 * 8], rax",
+            "mov [{} + 1 * 8], rbx",
+            "mov [{} + 2 * 8], rcx",
+            "mov [{} + 3 * 8], rdx",
+            "mov [{} + 4 * 8], rsi",
+            "mov [{} + 5 * 8], rdi",
+            "mov [{} + 6 * 8], rbp",
+            "mov [{} + 7 * 8], r8",
+            "mov [{} + 8 * 8], r9",
+            "mov [{} + 9 * 8], r10",
+            "mov [{} + 10 * 8], r11",
+            "mov [{} + 11 * 8], r12",
+            "mov [{} + 12 * 8], r13",
+            "mov [{} + 13 * 8], r14",
+            "mov [{} + 14 * 8], r15",
+            in(reg) &mut BSP_REGS,
+        );
+    }
+}
+
+pub fn restore_regs() {
+    unsafe {
+        asm!(
+            "mov rax, [{} + 0 * 8]",
+            "mov rbx, [{} + 1 * 8]",
+            "mov rcx, [{} + 2 * 8]",
+            "mov rdx, [{} + 3 * 8]",
+            "mov rsi, [{} + 4 * 8]",
+            "mov rdi, [{} + 5 * 8]",
+            "mov rbp, [{} + 6 * 8]",
+            "mov r8, [{} + 7 * 8]",
+            "mov r9, [{} + 8 * 8]",
+            "mov r10, [{} + 9 * 8]",
+            "mov r11, [{} + 10 * 8]",
+            "mov r12, [{} + 11 * 8]",
+            "mov r13, [{} + 12 * 8]",
+            "mov r14, [{} + 13 * 8]",
+            "mov r15, [{} + 14 * 8]",
+            in(reg) &BSP_REGS,
+        );
+    }
+}
```

Comment:
Not sure this will work, given this will give these functions rust's calling convention, and they are used in interrupts where there's another calling convention, using `global_asm!` here might be better.

Same applies to some other functions you added like this one

Reply:
Plus what I mentioned on discord about passing in(reg) &BSP_REGS and not sym BSP_REGS